subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/LegacyTestArtifactsProcessor.kt

```diff
@@ -0,0 +1,108 @@
+    private val gson: Gson,
+    private val testArtifactsUploader: TestArtifactsUploader,
+    private val retracer: ProguardRetracer,
+    private val timeProvider: TimeProvider
+) : TestArtifactsProcessor {
+
+    override fun process(
+        reportDir: File,
+        testStaticData: TestStaticData,
+        logcatBuffer: LogcatBuffer?
+    ): Result<AndroidTest> {
+
+        val scope = CoroutineScope(CoroutineName("test-artifacts-${testStaticData.name}") + Dispatchers.IO)
+
+        val reportJson = File(reportDir, TestArtifactsProcessor.REPORT_JSON_ARTIFACT)
+
+        return Result.tryCatch {
+            val testRuntimeData: TestRuntimeData = gson.fromJson<TestRuntimeDataPackage>(
+                FileReader(reportJson)
+            )
+
+            val isTestFailed = testRuntimeData.incident != null
+
+            runBlocking {
+                withContext(scope.coroutineContext) {
+                    AndroidTest.Completed.create(
+                        testStaticData = testStaticData,
+                        testRuntimeData = testRuntimeData,
+                        stdout = uploadLogcat(logcatBuffer?.getStdout(), isUploadNeeded = isTestFailed),
+                        stderr = uploadLogcat(logcatBuffer?.getStderr(), isUploadNeeded = isTestFailed)
+                    )
+                }
+            }
+        }
+    }
+
+    override fun processFailure(
+        throwable: Throwable,
+        testStaticData: TestStaticData,
+        logcatBuffer: LogcatBuffer?
+    ): Result<AndroidTest> {
+
+        val scope = CoroutineScope(CoroutineName("test-artifacts-failure-${testStaticData.name}") + Dispatchers.IO)
+
+        return runBlocking {
+            withContext(scope.coroutineContext) {
+                Result.Success(
+                    AndroidTest.Lost.fromTestMetadata(
+                        testStaticData,
+                        startTime = 0,
+                        lastSignalTime = 0,
+                        stdout = uploadLogcat(logcatBuffer?.getStdout(), isUploadNeeded = true),
+                        stderr = uploadLogcat(logcatBuffer?.getStderr(), isUploadNeeded = true),
+                        incident = Incident(
+                            type = Incident.Type.INFRASTRUCTURE_ERROR,
+                            timestamp = timeProvider.nowInSeconds(),
+                            trace = throwable.stackTraceToList(),
+                            chain = listOf(
+                                IncidentElement(
+                                    message = throwable.message ?: "no error message"
+                                )
+                            ),
+                            entryList = emptyList()
+                        )
+                    )
+                )
+            }
+        }
+    }
+
+    private suspend fun uploadLogcat(logcat: List<String>?, isUploadNeeded: Boolean): String {
+        return if (isUploadNeeded) {
+            if (logcat != null) {
```

Comment:
Nested if is harder to read.
Can we convert it to switch?

Reply:
done